src/Corax/Indexing/IndexWriter.cs

```diff
@@ -618,47 +650,55 @@ private bool TryDeleteEntry(Slice termSlice, out long entryId)
+                    RecordAndPrepareDocumentsIdsForDeletion(containerId, out var setsAreDisjoint, out var isSingleDocument);
+                    entryId = isSingleDocument ? _entriesToDelete[0] : Constants.IndexSearcher.InvalidId;                        
+                    ProcessCurrentDeletes();
+                    
                     Debug.Assert(setsAreDisjoint, "The set scheduled for deletion shares common elements with the posting list. This should not be possible here.");
-                    return result;
+                    return isSingleDocument;
                 }
-                entryId = -1;
+                entryId = Constants.IndexSearcher.InvalidId;
                 return false;
             }
 
             FlushBatch();
             return TryDeleteEntry(termSlice, out entryId);
         }
 
-        public void DeleteByPrefix(ReadOnlySpan<byte> term)
+        public void DeleteByPrefix(ReadOnlySpan<byte> prefix)
         {
-            using var __ = Slice.From(_transaction.Allocator, term, ByteStringType.Immutable, out var termSlice);
-
-            if (_indexedEntries.Any(id => SliceComparer.StartWith(id, termSlice)) == false)
+            using var __ = Slice.From(_transaction.Allocator, prefix, ByteStringType.Immutable, out var prefixSlice);
+            var hasPrefixInCurrentlyIndexedEntries = _indexedEntries.Any(id => SliceComparer.StartWith(id, prefixSlice));
+            var requiresFlushingBatch = hasPrefixInCurrentlyIndexedEntries;
+            
+            if (hasPrefixInCurrentlyIndexedEntries == false)
             {
-                  var fieldsTree = _fieldsTree.CompactTreeFor(_fieldsMapping.GetByFieldId(Constants.IndexWriter.PrimaryKeyFieldId).FieldName);
-
-                var iterator = fieldsTree.Iterate();
-                iterator.Seek(termSlice);
+                var primaryKeyTree = _fieldsTree.CompactTreeFor(_fieldsMapping.GetByFieldId(Constants.IndexWriter.PrimaryKeyFieldId).FieldName);
+                var treeIterator = primaryKeyTree.Iterate();
+                treeIterator.Seek(prefixSlice);
 
-                while (iterator.MoveNext(out var key, out var id, out _))
+                while (requiresFlushingBatch == false && treeIterator.MoveNext(out var currentKey, out var postingListId, out _))
                 {
-                    if (key.Decoded().StartsWith(termSlice) == false)
-                        continue;
+                    // Since we're seeking, we will have a key that has exactly the same prefix or prefix + 1 (in terms of order).
+                    if (currentKey.Decoded().StartsWith(prefixSlice) == false)
+                        break;
 
-                    RecordDeletion(id, out _, out _);
+                    RecordAndPrepareDocumentsIdsForDeletion(postingListId, out bool setsAreDisjoint, out _);
+                    ProcessCurrentDeletes();
+                    requiresFlushingBatch |= setsAreDisjoint == false;
```

Comment:
Do you need `|=`? Once it's set to `true` then we exit the loop anyway

Reply:
You're right, changed.